### PR TITLE
i#6662 public traces: add signal and uncompleted instr invariants

### PR DESCRIPTION
**Pull request by edeiana**
Public traces should never have TRACE_MARKER_TYPE_SIGNAL_NUMBER markers.
Temporarily, we also disallow TRACE_MARKER_TYPE_UNCOMPLETED_INSTRUCTION
markers until we can handle their partial encoding. Eventually we want to preserve them
(xref: #7155).
We add two invariants in invariant_checker to detect the presence of these
markers in REGDEPS traces and raise an invariant error if necessary.

Updated unit tests and end-to-end tests.

Issue #6662 #7155

### CODE REVIEW

File: clients/drcachesim/tests/invariant_checker_test.cpp [hunk 1]
```diff
@@ -3581,6 +3581,48 @@ check_regdeps(void)
...
+                { "OFFLINE_FILE_TYPE_ARCH_REGDEPS traces cannot have "
+                  "TRACE_MARKER_TYPE_SIGNAL_NUMBER markers",
+                  /*tid=*/TID_A,
+                  /*ref_ordinal=*/4, /*last_timestamp=*/0,
+                  /*instrs_since_last_timestamp=*/0 },
+                "Failed to catch non-allowed TRACE_MARKER_TYPE_SIGNAL_NUMBER marker"))
+            return false;
+    }
+
+    // Incorrect: TRACE_MARKER_TYPE_UNCOMPLETED_INSTRUCTION not allowed.
```
Comment (derekbruening) [inline clients/drcachesim/tests/invariant_checker_test.cpp:3605]: Add a XXX i# comment about updating/removing this once we handle these.

**Review by derekbruening — APPROVED** [review]

File: suite/tests/CMakeLists.txt [hunk 2]
```diff
@@ -4849,10 +4849,11 @@ if (BUILD_CLIENTS)
           # -filter_marker_types 19,25,27,28,30 (which correspond to
           # TRACE_MARKER_TYPE_SYSCALL_IDX, TRACE_MARKER_TYPE_SYSCALL,
           # TRACE_MARKER_TYPE_SYSCALL_TRACE_START, TRACE_MARKER_TYPE_SYSCALL_TRACE_END,
-          # TRACE_MARKER_TYPE_SYSCALL_FAILED), and -filter_modify_marker_value 3,-1
-          # (which changes the value of TRACE_MARKER_TYPE_CPU_ID == 3 to
-          # INVALID_CPU_MARKER_VALUE == (uintptr_t)-1).
-          "${drcachesim_path}@-simulator_type@record_filter@-filter_encodings2regdeps@-indir@${testname}.${ci_shared_app}.*.dir/trace@-outdir@${testname}.filtered.dir@-filter_marker_types@19,25,27,28,30@-filter_keep_func_ids@4294967498@-filter_modify_marker_value@3,-1"
+          # TRACE_MARKER_TYPE_SYSCALL_FAILED,  TRACE_MARKER_TYPE_SIGNAL_NUMBER,
+          # TRACE_MARKER_TYPE_UNCOMPLETED_INSTRUCTION), and
```
[the remark this answers is not in the document]
Comment (edeiana) [reply suite/tests/CMakeLists.txt:4853]: Done.

File: suite/tests/CMakeLists.txt [hunk 3]
```diff
@@ -4849,10 +4849,11 @@ if (BUILD_CLIENTS)
           # -filter_marker_types 19,25,27,28,30 (which correspond to
```
[the remark this answers is not in the document]
Comment (edeiana) [reply suite/tests/CMakeLists.txt:4849]: Comment fixed.

File: suite/tests/CMakeLists.txt [hunk 4]
```diff
@@ -4866,13 +4867,14 @@ if (BUILD_CLIENTS)
...
           # TRACE_MARKER_TYPE_SYSCALL_IDX, TRACE_MARKER_TYPE_SYSCALL,
           # TRACE_MARKER_TYPE_SYSCALL_TRACE_START, TRACE_MARKER_TYPE_SYSCALL_TRACE_END,
-          # TRACE_MARKER_TYPE_SYSCALL_FAILED), and -filter_modify_marker_value 3,-1
-          # (which changes the value of TRACE_MARKER_TYPE_CPU_ID == 3 to
-          # INVALID_CPU_MARKER_VALUE == (uintptr_t)-1).
-          "${drcachesim_path}@-simulator_type@record_filter@-filter_encodings2regdeps@-indir@${testname}.${kernel_xfer_app}.*.dir/trace@-outdir@${testname}.filtered.dir@-filter_marker_types@19,25,27,28,30@-filter_keep_func_ids@4294967498@-filter_modify_marker_value@3,-1"
+          # TRACE_MARKER_TYPE_SYSCALL_FAILED, TRACE_MARKER_TYPE_SIGNAL_NUMBER,
+          # TRACE_MARKER_TYPE_UNCOMPLETED_INSTRUCTION), and
+          # -filter_modify_marker_value 3,-1 (which changes the value of
+          # TRACE_MARKER_TYPE_CPU_ID == 3 to INVALID_CPU_MARKER_VALUE == (uintptr_t)-1).
```
[the remark this answers is not in the document]
Comment (edeiana) [reply suite/tests/CMakeLists.txt:4876]: Unified at the top.

File: clients/drcachesim/tools/invariant_checker.cpp [hunk 5]
```diff
@@ -1707,6 +1707,16 @@ invariant_checker_t::check_regdeps_invariants(per_shard_t *shard, const memref_t
                          "was not built on a Linux platform.\n";
 #endif
         } break;
+        case TRACE_MARKER_TYPE_SIGNAL_NUMBER:
+            report_if_false(shard, false,
+                            "OFFLINE_FILE_TYPE_ARCH_REGDEPS traces cannot have "
+                            "TRACE_MARKER_TYPE_SIGNAL_NUMBER markers");
+            break;
+        case TRACE_MARKER_TYPE_UNCOMPLETED_INSTRUCTION:
```
[the remark this answers is not in the document]
Comment (edeiana) [reply clients/drcachesim/tools/invariant_checker.cpp:1715]: Done.

File: clients/drcachesim/tests/invariant_checker_test.cpp (hunk 1 above)
Comment (edeiana) [reply clients/drcachesim/tests/invariant_checker_test.cpp:3605]: Done.

**Pull request merged**